server/src/test/kotlin/fi/oph/kitu/KituApplicationTests.kt

```diff
@@ -1,11 +1,22 @@
 package fi.oph.kitu
 
-import fi.oph.kitu.test.DBFixture
 import org.junit.jupiter.api.Test
 import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.boot.testcontainers.service.connection.ServiceConnection
+import org.testcontainers.containers.PostgreSQLContainer
+import org.testcontainers.junit.jupiter.Container
+import org.testcontainers.junit.jupiter.Testcontainers
 
 @SpringBootTest
-class KituApplicationTests : DBFixture() {
+@Testcontainers
+class KituApplicationTests {
+    companion object {
+        @JvmStatic
+        @Container
+        @ServiceConnection
+        val postgres = PostgreSQLContainer("postgres:16")
+    }
```

Comment:
Tämä koko hässäkkä olisi Javassa

```java
@Container
@ServiceConnection
static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:16");
```

Sentään Kotlinissa ei tarvitse toistaa tyyppiä!

Reply:
Onko mahdollista/järkevää, että tämä laitettaisiin johonkin base Classiin, esim IntegrationTestsBase, jonka meidän testit peris? Vähentäis koodimäärää.

Reply:
Kokeilin sitä: ensimmäisen testin jälkeen kontti sammuu ja muut testit eivät saa siihen yhteyttä enää. En löytänyt dokumentaatiota miten sen saisi toimimaan.

Reply:
> Onko mahdollista/järkevää, että tämä laitettaisiin johonkin base Classiin

Ongelma taitaa olla, että `postgres` instanssin on oltava `static`, jotta staattiset `@BeforeAll`/`@AfterAll` -metodit pääsevät käynnistämään/sammuttamaan kannan kontin.

> ensimmäisen testin jälkeen kontti sammuu ja muut testit eivät saa siihen yhteyttä enää

Villi veikkaus, mutta tämä voisi johtua siitä, että staattiset kentät/metodit ja companion objektit eivät periydy nätisti (lue: eivät periydy ollenkaan), jolloin käytännössä vain abstraktilla luokalla on yksi staattinen `postgres`-instanssi jota kaikki testit käyttävät kilpaa. Eli syntyy kilpailutilanne ja ensimmäisenä loppuun suoritettu testi sammuttaa kannan. Tai jotain vastaavaa.

Jos näin on, yksi mahdollinen tapa kiertää ongelma saattaisi olla tekemällä `postgres` instanssista ei-staattinen. Tämä on hieman haastavaa: jotta `postgres` ei olisi staattinen, myös `@BeforeAll` ja `@AfterAll` -metodien (lifecycle method) tulee olla ei-staattisia. JUnit5 oletuskonfiguraatiolla tämä ei taida olla mahdollista. Test suiten luokasta luodaan uusi instanssi per-testitapaus ja lifecycle metodien staattisuus taitaa olla pakotettu konventio.

Tätä on kuitenkin mahdollista muuttaa. Test Instance Lifecycle on mahdollista asettaa `PER_METHOD` sijaan `PER_CLASS`, jolloin test suiten instanssia uudelleenkäytetään läpi luokassa määriteltyjen testitapausten ([JUnit dokumentaatio](https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle-changing-default)). Näin konfiguroituna, `@BeforeAll` ja `@AfterAll` ei tarvitse olla staattisia ja nämä voivat näinollen vapaasti käyttää luokan ei-staattisia kenttiä.

Pidän `PER_METHOD` -> `PER_CLASS` konfiguraatiomuutosta kuitenkin jokseenkin huonona ideana. Luokan tilan jakaminen testisuorituksesta toiseen johtaa herkästi ties mihin ikäviin sivuvaikutuksiin. On hirveän houkuttelevaa siirtää useassa testitapauksessa käytettäviä asioita testiluokan kentiksi, jolloin `PER_CLASS` -lifecyclen ollessa käytössä, on riskinä että syntyy testitapausten välisiä riippuvuuksia. Vastaavista syistä myös esim. testien rinnakkaistaminen voi hankaloitua huomattavasti. En kuitenkaan tiedä kuinka todellisia nuo riskit/uhkakuvat ovat.

Reply:
Suorastaan mainiota kaivelua @Kailari. Kiitos. 🙇